mountpoint-s3-crt/src/s3/pool.rs

```diff
@@ -0,0 +1,432 @@
+        // SAFETY: We're setting up the struct to be self-referential, and we're not moving out
+        // of the struct, so the unchecked deref of the pinned pointer is okay.
+        unsafe {
+            let pool_ref = Pin::get_unchecked_mut(Pin::as_mut(&mut crt_pool));
+            pool_ref.inner.vtable = &raw mut pool_ref.pool_vtable;
+            pool_ref.inner.impl_ = pool_ref as *mut CrtBufferPool<Pool> as *mut libc::c_void;
+            aws_ref_count_init(
+                &mut pool_ref.inner.ref_count,
+                &mut pool_ref.inner as *mut aws_s3_buffer_pool as *mut libc::c_void,
+                Some(pool_destroy::<Pool>),
+            );
+        }
+
+        crt_pool
+    }
+
+    /// Leak a pinned instance and returns a raw pointer.
+    ///
+    /// # Safety
+    /// The returned pointer must eventually be passed to [from_raw] and can
+    /// additionally only used in [ref_from_raw].
+    unsafe fn leak(self: Pin<Box<Self>>) -> *mut aws_s3_buffer_pool {
+        // SAFETY: the resulting pointer will be only used in `pool_reserve`, `pool_trim`, and `pool_destroy`.
+        let pool = Box::leak(unsafe { Pin::into_inner_unchecked(self) });
+        &raw mut pool.inner
+    }
+
+    /// Returns a reference to original instance from a raw pointer.
+    ///
+    /// # Safety
+    /// The raw pointer must have been obtained through [leak()].
+    unsafe fn ref_from_raw(pool: &*mut aws_s3_buffer_pool) -> &Self {
+        // SAFETY: `pool` points to the `inner` field of a pinned instance.
+        unsafe {
+            let impl_ptr = (**pool).impl_;
+            &*(impl_ptr as *mut Self)
+        }
+    }
+
+    /// Re-constructs the original pinned instance from a raw pointer.
+    ///
+    /// # Safety
+    /// The raw pointer must have been obtained through [leak()].
+    unsafe fn from_raw(pool: *mut aws_s3_buffer_pool) -> Pin<Box<Self>> {
+        // SAFETY: `pool` points to the `inner` field of a pinned instance.
+        unsafe { Pin::new_unchecked(Box::from_raw((*pool).impl_ as *mut Self)) }
+    }
+
+    fn trim(&self) {
+        self.pool.trim();
+    }
+
+    fn reserve(&self, size: usize) -> CrtTicketFuture {
+        let future = CrtTicketFuture::new(&self.allocator);
+
+        // Get a buffer from the pool, build its ticket, and immediately fullfil the future.
+        // This will likely change later, when we make the method on the pool async.
+        let buffer = self.pool.get_buffer(size);
+        let ticket = self.make_ticket(buffer);
+        future.set(ticket);
+
+        future
+    }
+
+    fn make_ticket(&self, buffer: Pool::Buffer) -> Pin<Box<CrtTicket<Pool::Buffer>>> {
+        // Set up the vtable by pointing into the pool pinned address, but leave `impl_`
+        // and `ref_count` to be initialized after pinning the ticket.
+        let mut ticket = Box::pin(CrtTicket {
+            inner: aws_s3_buffer_ticket {
+                vtable: (&raw const self.ticket_vtable).cast_mut(),
```

Comment:
may `self` (and `aws_s3_buffer_ticket->vtable`) be destroyed before the returned `aws_s3_buffer_ticket`?